ouroboros-network/src/Ouroboros/Network/Mux.hs

```diff
@@ -59,12 +60,21 @@ muxJobs (MiniProtocolDescriptions udesc) wfn rfn sdufn = do
         b <- atomically $ newTBQueue 2
         return $ M.insert (p, ModeInitiator) a $ M.insert (p, ModeResponder) b t
 
-    mpsJob pmss mpd = do
+    mpsJob cnt pmss mpd = do
         w_i <- atomically newEmptyTMVar
         w_r <- atomically newEmptyTMVar
 
-        return [ mpdInitiator mpd $ muxChannel pmss (mpdId mpd) ModeInitiator w_i
-               , mpdResponder mpd $ muxChannel pmss (mpdId mpd) ModeResponder w_r]
+        return [ mpdInitiator mpd (muxChannel pmss (mpdId mpd) ModeInitiator w_i cnt)
+                     >> mpsJobExit cnt
+               , mpdResponder mpd (muxChannel pmss (mpdId mpd) ModeResponder w_r cnt)
+                     >> mpsJobExit cnt
+               ]
+
+    -- cnt represent the number of SDUs that are queued but not yet sent.
+    -- job threads will be prevented from exiting until all SDUs have been transmitted.
```

Comment:
I'm not sure I totally get this explanation for why we have the counter, probably because I don't yet get the wider context.

Reply:
At the moment there are two ways to tear down a Mux Bearer, throw an exception or have a miniprotocol job thread exit. There is a risk that the application would send its last message and then directly decide it was done and exit. That could result in that the Mux Bearer was torn down before the `mux` thread had time to send the final message which would leave the other side waiting forever (MuxSTMCtx case) or an abnormal exit (Socket).

Reply:
Would an alternative be to have the mini-protocol job close its end of the channel, and then the mux eventually reads all the outbound data and sees that the channel has closed?

Reply:
Not with the current implementation since it uses `waitAnyCatchCancel`, when a mini-protcol job thread exits everything goes down.